### PR TITLE
v3: Move DS Template storage setting

### PR DESCRIPTION
**Pull request by JillieBeanSim**
## Proposed changes

https://github.com/zowe/zowe-explorer-vscode/issues/2345

To Do:
- [x] move setting out of ds history to "zowe.ds.templates"
- [x] auto migration of setting from history to new, User setting only with migration as it was part of history
- [x] add unit tests
- [x] update v3 changes doc

If no workspace open when saving templates auto save to User Settings.
If workspace open when saving templates, user if asked if they would like to save to User or Workspace Settings.

**See Comment Below about migration testing**

example history templates:
```
"templates": [
      {
        "newtst": {
          "alcunit": "CYL",
          "blksize": 6160,
          "dirblk": 10,
          "dsorg": "PO",
          "primary": 1,
          "recfm": "FB",
          "lrecl": 80
        }
      },
      {
        "newtst2": {
          "alcunit": "CYL",
          "blksize": 6160,
          "dirblk": 5,
          "dsorg": "PO",
          "primary": 1,
          "recfm": "FB",
          "lrecl": 80
        }
      }
  ]
```

## Release Notes

Milestone: 3.0.0

Changelog:

- ZE API:
  - **Breaking:** Added return type of `Promise<void>` to `IZoweTree.addDsTemplate`. [#2783](https://github.com/zowe/vscode-extension-for-zowe/pull/2783).
- ZE:
  - **BREAKING** Moved data set templates out of data set history settings into new setting "zowe.ds.templates". [#2345](https://github.com/zowe/zowe-explorer-vscode/issues/2345)

## Types of changes

What types of changes does your code introduce to Zowe Explorer?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer_

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/zowe-explorer-vscode/wiki/Best-Practices:-Contributor-Guidance) wiki
- [x] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [x] PR Description is included
- [x] gif or screenshot is included if visual changes are made
- [x] `yarn workspace vscode-extension-for-zowe vscode:prepublish` has been executed
- [x] All checks have passed (DCO, Jenkins and Code Coverage)
- [x] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [x] There is coverage for the code that I have added
- [x] I have tested it manually and there are no regressions found
- [x] I have added necessary documentation (if appropriate)
- [ ] Any PR dependencies have been merged and published (if appropriate)

## Further comments

https://github.com/zowe/zowe-explorer-vscode/assets/49491949/cc42d4f4-f0cc-426e-9056-c67721adf2ea

### CODE REVIEW

File: packages/zowe-explorer/src/configuration/SettingsConfig.ts
Comment (JillieBeanSim) [inline packages/zowe-explorer/src/configuration/SettingsConfig.ts:89]: Attn all testers!!! 👋🏻 

Uncomment this line for testing the migration steps especially if you have already been migrated to internal storage for v3.

File: packages/zowe-explorer-api/CHANGELOG.md
Comment (t1m0thyj) [inline packages/zowe-explorer-api/CHANGELOG.md:17]: Looks like the PR links may have gotten swapped

File: packages/zowe-explorer/src/configuration/Constants.ts
Comment (t1m0thyj) [inline packages/zowe-explorer/src/configuration/Constants.ts:79]: Should we mark other constants in this file as readonly too?

**Review by t1m0thyj — COMMENTED** [review]
LGTM so far, thanks @JillieBeanSim! Left a few comments and will approve after testing the changes 😋

File: packages/zowe-explorer/src/configuration/Constants.ts
Comment (JillieBeanSim) [reply packages/zowe-explorer/src/configuration/Constants.ts:79]: we can, I got flagged for my addition by sonarcloud.

File: packages/zowe-explorer/__tests__/__unit__/trees/dataset/DatasetTree.unit.test.ts
Comment (t1m0thyj) [inline packages/zowe-explorer/__tests__/__unit__/trees/dataset/DatasetTree.unit.test.ts:3025]: In addition to testing that an info message is shown, should we test that a function to update the setting was called? Either `DataSetTemplates.updateDsTemplateSetting` or `SettingsConfig.setDirectValue`.

File: packages/zowe-explorer/src/trees/dataset/DatasetTemplates.ts
Comment (t1m0thyj) [inline packages/zowe-explorer/src/trees/dataset/DatasetTemplates.ts:85]: Can we reuse `getDsTemplates` or `SettingsConfig.getDirectValue` here to reduce code duplication?

**Review by t1m0thyj — APPROVED** [review]
LGTM, thanks @JillieBeanSim! Left a few comments but nothing that needs to be addressed before merge

File: packages/zowe-explorer/src/trees/dataset/DatasetTemplates.ts
Comment (JillieBeanSim) [reply packages/zowe-explorer/src/trees/dataset/DatasetTemplates.ts:85]: this line calls inspect vs get like suggested methods for the configuration to allow check for local vs global

**Review by t1m0thyj — APPROVED** [review]
LGTM, thanks @JillieBeanSim!

**Review by zFernand0 — APPROVED** [review]
Thanks for the clarification 😊 
LGTM! 😋

**Pull request merged**
